relays/client-substrate/src/client/caching.rs

```diff
@@ -175,41 +212,45 @@ impl<C: Chain, B: Client<C>> Client<C> for CachingClient<C, B> {
 	}
 
 	async fn best_finalized_header_hash(&self) -> Result<HashOf<C>> {
-		// TODO: after https://github.com/paritytech/parity-bridges-common/issues/2074 we may
-		// use single-value-cache here, but for now let's just call the backend
-		self.backend.best_finalized_header_hash().await
+		self.read_header_from_background(
+			&self.data.best_finalized_header,
+			self.backend.best_finalized_header(),
+		)
+		.and_then(|h| futures::future::ready(Ok(h.hash())))
+		.await
 	}
 
 	async fn best_header(&self) -> Result<HeaderOf<C>> {
-		// TODO: if after https://github.com/paritytech/parity-bridges-common/issues/2074 we'll
-		// be using subscriptions to get best blocks, we may use single-value-cache here, but for
-		// now let's just call the backend
-		self.backend.best_header().await
+		self.read_header_from_background(&self.data.best_header, self.backend.best_header())
+			.await
 	}
 
-	async fn subscribe_grandpa_finality_justifications(&self) -> Result<Subscription<Bytes>>
-	where
-		C: ChainWithGrandpa,
-	{
-		let mut grandpa_justifications = self.data.grandpa_justifications.lock().await;
-		if let Some(ref grandpa_justifications) = *grandpa_justifications {
-			grandpa_justifications.subscribe().await
-		} else {
-			let subscription = self.backend.subscribe_grandpa_finality_justifications().await?;
-			*grandpa_justifications = Some(subscription.factory());
-			Ok(subscription)
-		}
+	async fn subscribe_best_headers(&self) -> Result<Subscription<HeaderOf<C>>> {
+		// we may share the sunbscription here, but atm there's no callers of this method
+		self.backend.subscribe_best_headers().await
+	}
+
+	async fn subscribe_finalized_headers(&self) -> Result<Subscription<HeaderOf<C>>> {
+		// we may share the sunbscription here, but atm there's no callers of this method
+		self.backend.subscribe_finalized_headers().await
+	}
+
+	async fn subscribe_grandpa_finality_justifications(&self) -> Result<Subscription<Bytes>> {
+		let (result_sender, result_receiver) = bounded(1);
+		self.to_background_sender
+			.read()
+			.await
+			.try_send(ToBackground::SubscribeGrandpaJustifications(result_sender))?;
+		result_receiver.recv().await.map_err(Into::into)
```

Comment:
Do we need to do this on the background worker ? Can we just do `self.backend.subscribe_grandpa_finality_justifications().await` here and also cache it ?